### PR TITLE
Limited LPPE support for JDBC connections

### LINKED ISSUE #2355: RFE : Add limited support for LPPE to JDBC Authentication
Currently the LDAP authentication mechanisms support LPPE combined with the password management functionality. As DB password management is now in a basic form incorporated into the password management module, it would be nice to add limited LPPE support for DB connections.

The basic requirements would be, 
expired password : forcing a user to change their password
disabled account : displaying a warning if an account is disabled

Others options may be further added

### PR DESCRIPTION
**Pull request by johnlister**
Closes #2355 

- This adds support for expired and disabled fields in the query and queryAndEncode database authentication mechanisms. Simulates LPPE with LDAP authentication handlers
- Test cases added, documentation updated to reflect new properties
- no known side effects

### CODE REVIEW

**Comment by johnlister** [comment]
Only Query and QueryAndEncode supported as doesn't seem appropriate with bind. Could be implemented with more work for the Search handler, but as this matches against any matching user, could cause conflicts

File: core/cas-server-core-configuration/src/main/java/org/apereo/cas/configuration/model/support/jdbc/JdbcAuthenticationProperties.java [hunk 1]
```diff
@@ -328,6 +357,22 @@ public void setSaltFieldName(final String saltFieldName) {
             this.saltFieldName = saltFieldName;
         }
 
+        public String getExpiredFieldName() {
```
Comment (mmoayyed) [inline core/cas-server-core-configuration/src/main/java/org/apereo/cas/configuration/model/support/jdbc/JdbcAuthenticationProperties.java]: These seem to duplicate the existing fields from the parent class?

Comment (mmoayyed) [inline core/cas-server-core-configuration/src/main/java/org/apereo/cas/configuration/model/support/jdbc/JdbcAuthenticationProperties.java]: Docs also need to be updated to ensure all handlers reference the same fields.

Comment (mmoayyed) [inline core/cas-server-core-configuration/src/main/java/org/apereo/cas/configuration/model/support/jdbc/JdbcAuthenticationProperties.java]: Also please update the docs in general to explain that password policy is possible now with JDBC to a limited extent.

And fix the build failures.

**Comment by mmoayyed** [comment]
hey @johnlister thanks very much again. I realize this pull might still need some work, but I am going to merge and carry on offline. 
Once the changes show up in snapshots, etc please review and make other contributions/modifications as you see fit.

**Comment by johnlister** [comment]
Hi, apologies I added the changes you requested, but I've struggled to get the build to run successfully locally. If I follow the instructions it builds fine, but if I use the build commands on travis I get a number of failures - either findBugs or checkStyle. These are in completely unrelated modules.

Clearly my local setup is broken somewhere...

Thanks for picking it up - any questions about the behaviour let me know

**Comment by mmoayyed** [comment]
All good. Thank you! 

I was just chatting with someone on gitter yesterday about this improvement and I think we are going to enhance it further. Thanks for getting the ball rolling.

I am also going to add this improvement to the changelog on the apereo blog.

**Comment by johnlister** [comment]
I totally agree, I put in the basics for what I needed currently, but we have internally been talking about adding allowed time fields, password expiry dates, etc I stopped short of putting those in now as would require more work to generalise the password management side as well. I look forward to seeing what you come up with

**Comment by mmoayyed** [comment]
Very cool. Going to ping @bbooth here to see if we can come up with a common/shared mechanism to enforce policies. 

Blog updated https://apereo.github.io/2017/02/24/rc2-release/#database-password-policy-enforcement

**Comment by johnlister** [comment]
Definitely, it should be straightforward to implement, just kicking myself I can't get the main branch to build locally...

**Comment by mmoayyed** [comment]
How do you build and what does it tell you?

**Comment by johnlister** [comment]
Using the standard instructions:
`gradlew build --parallel -x test -x javadoc -DskipCheckstyle=true -DskipFindbugs=true`

works fine, but travis does:

`gradlew clean bootRepackage install --parallel -x test --stacktrace`
followed by
`gradlew checkstyleTest test --stacktrace --parallel' 

It used to fail with checkStyle issues on mongo related modules, but just done it now and the ldap tests are failing - presumably I need a server to connect to?

**Comment by mmoayyed** [comment]
No. CAS ships with an embedded LDAP server for testing. [That's how Travis passes tests] There are no external setups. You might have something that is preventing the server to start up. 

You should also remove all `-x test -x javadoc -DskipCheckstyle=true -DskipFindbugs=true` prior to shooting a pull request and building. Your build must be consistent with Travis as much as possible. That's how snapshots get build, and if that passes, then we know it's all good.

**Pull request merged**